src/sentry/consumers/dlq.py

```diff
@@ -0,0 +1,146 @@
+
+    def __init__(
+        self,
+        producers: Mapping[RejectReason, KafkaDlqProducer | None],
+    ) -> None:
+        self.producers = producers
+
+    def produce(
+        self,
+        value: BrokerValue[KafkaPayload],
+        reason: str | None = None,
+    ) -> Future[BrokerValue[KafkaPayload]]:
+
+        reject_reason = RejectReason(reason) if reason else RejectReason.INVALID
+        producer = self.producers.get(reject_reason)
+
+        if producer:
+            return producer.produce(value)
+        else:
+            # No DLQ producer configured for the reason.
+            logger.error("No DLQ producer configured for reason %s", reason)
+            future: Future[BrokerValue[KafkaPayload]] = Future()
+            future.set_running_or_notify_cancel()
+            future.set_result(value)
+            return future
+
+
+def _get_dlq_producer(topic: Topic | None) -> KafkaDlqProducer | None:
+    if topic is None:
+        return None
+
+    topic_defn = get_topic_definition(topic)
+    config = get_kafka_producer_cluster_options(topic_defn["cluster"])
+    real_topic = topic_defn["real_topic_name"]
+    return KafkaDlqProducer(KafkaProducer(config), ArroyoTopic(real_topic))
+
+
+def maybe_build_dlq_producer(
+    dlq_topic: Topic | None,
+    stale_topic: Topic | None,
+) -> MultipleDestinationDlqProducer | None:
+    if dlq_topic is None and stale_topic is None:
+        return None
+
+    producers = {
+        RejectReason.INVALID: _get_dlq_producer(dlq_topic),
+        RejectReason.STALE: _get_dlq_producer(stale_topic),
+    }
+
+    return MultipleDestinationDlqProducer(producers)
+
+
+class DlqStaleMessages(ProcessingStrategy[KafkaPayload]):
+    def __init__(
+        self,
+        stale_threshold_sec: int,
+        next_step: ProcessingStrategy[KafkaPayload | FilteredPayload],
+    ) -> None:
+        self.stale_threshold_sec = stale_threshold_sec
+        self.next_step = next_step
+
+        # A filtered message is created so we commit periodically if all are stale.
+        self.last_forwarded_offsets = time.time()
+        self.offsets_to_forward: MutableMapping[Partition, int] = {}
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        min_accepted_timestamp = datetime.now(timezone.utc) - timedelta(
+            seconds=self.stale_threshold_sec
+        )
+
+        if isinstance(message.value, BrokerValue):
+            if message.value.timestamp < min_accepted_timestamp:
+                self.offsets_to_forward[message.value.partition] = message.value.next_offset
+                raise InvalidMessage(
+                    message.value.partition, message.value.offset, reason=RejectReason.STALE.value
+                )
+
+        if self.offsets_to_forward and time.time() > self.last_forwarded_offsets + 1:
+            filtered_message = Message(Value(FILTERED_PAYLOAD, self.offsets_to_forward))
+            self.offsets_to_forward = {}
+            self.next_step.submit(filtered_message)
```

Comment:
TODO: this needs to be fixed/moved or we won't get here if we we have only invalid messages